README.md
# aurelia-event-aggregator

[![npm Version](https://img.shields.io/npm/v/aurelia-event-aggregator.svg)](https://www.npmjs.com/package/aurelia-event-aggregator)
[![ZenHub](https://raw.githubusercontent.com/ZenHubIO/support/master/zenhub-badge.png)](https://zenhub.io)
[![Join the chat at https://gitter.im/aurelia/discuss](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/aurelia/discuss?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![CircleCI](https://circleci.com/gh/aurelia/event-aggregator.svg?style=shield)](https://circleci.com/gh/aurelia/event-aggregator)

This library is part of the [Aurelia](http://www.aurelia.io/) platform and contains a lightweight pub/sub messaging system for app-wide or per-object loosely coupled events.

> To keep up to date on [Aurelia](http://www.aurelia.io/), please visit and subscribe to [the official blog](http://blog.aurelia.io/) and [our email list](http://eepurl.com/ces50j). We also invite you to [follow us on twitter](https://twitter.com/aureliaeffect). If you have questions, please [join our community on Gitter](https://gitter.im/aurelia/discuss) or use [stack overflow](http://stackoverflow.com/search?q=aurelia). Documentation can be found [in our developer hub](http://aurelia.io/hub.html). If you would like to have deeper insight into our development process, please install the [ZenHub](https://zenhub.io) Chrome or Firefox Extension and visit any of our repository's boards.

## Platform Support

This library can be used in the **browser** as well as on the **server**.

## Building The Code

To build the code, follow these steps.

1. Ensure that [NodeJS](http://nodejs.org/) is installed. This provides the platform on which the build tooling runs.
2. From the project folder, execute the following command:

	```shell
	npm install
	```
4. To build the code, you can now run:

	```shell
	npm run build
	```
5. You will find the compiled code in the `dist` folder, available in six module formats: AMD, CommonJS and ES2015, ES2017, native modules, and System.

6. See `package.json` scripts section for other tasks related to generating the docs and linting.

## Running The Tests

To run the unit tests, first ensure that you have followed the steps above in order to install all dependencies and successfully build the library. Once you have done that, proceed with these additional steps:

```shell
npm test
```


doc/CHANGELOG.md
<a name="1.0.4"></a>
## [1.0.4](https://github.com/aurelia/event-aggregator/compare/1.0.2...1.0.4) (2019-03-27)


### Bug Fixes

* **all:** change es2015 back to native-modules ([38c3b31](https://github.com/aurelia/event-aggregator/commit/38c3b31))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/aurelia/event-aggregator/compare/1.0.2...1.0.3) (2019-02-04)


### Bug Fixes

* **all:** change es2015 back to native-modules ([38c3b31](https://github.com/aurelia/event-aggregator/commit/38c3b31))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/aurelia/event-aggregator/compare/1.0.1...1.0.2) (2019-01-18)

* Add module field to package.json

<a name="1.0.1"></a>
## [1.0.1](https://github.com/aurelia/event-aggregator/compare/1.0.0...v1.0.1) (2016-12-03)


### Bug Fixes

* **index:** handle errors per callback invocation ([dc7819f](https://github.com/aurelia/event-aggregator/commit/dc7819f))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/aurelia/event-aggregator/compare/1.0.0-rc.1.0.0...v1.0.0) (2016-07-27)



<a name="1.0.0-rc.1.0.0"></a>
# [1.0.0-rc.1.0.0](https://github.com/aurelia/event-aggregator/compare/1.0.0-beta.2.0.1...v1.0.0-rc.1.0.0) (2016-06-22)



### 1.0.0-beta.1.2.1 (2016-05-10)


### 1.0.0-beta.1.2.0 (2016-03-22)

* Update to Babel 6

### 1.0.0-beta.1.1.1 (2016-02-08)


#### Bug Fixes

* **index:** add some error checking to guard args ([e1661a00](http://github.com/aurelia/event-aggregator/commit/e1661a00ff9894891a59cd8dfd3f5d83964de158), closes [#15](http://github.com/aurelia/event-aggregator/issues/15))


### 1.0.0-beta.1.1.0 (2016-01-29)


#### Features

* **all:** update jspm meta; core-js; aurelia deps ([05e6674d](http://github.com/aurelia/event-aggregator/commit/05e6674d13d393e6813dda656a8e79ba75b85aa8))


### 1.0.0-beta.1 (2015-11-16)


## 0.10.0 (2015-11-10)


## 0.9.0 (2015-10-13)


#### Bug Fixes

* **all:**
  * update to latest plugin api ([2a17472d](http://github.com/aurelia/event-aggregator/commit/2a17472dba61e6d6f364207d2ba438e65a97ed87))
  * upgrade to latest compiler ([ae84b313](http://github.com/aurelia/event-aggregator/commit/ae84b3139c1cc906614a904d78935e06a7959fde))
* **build:**
  * update linting, testing and tools ([172366a4](http://github.com/aurelia/event-aggregator/commit/172366a4e3ce6b97c16fc0f6af96e2e32114ba2e))
  * add missing bower bump ([31130469](http://github.com/aurelia/event-aggregator/commit/31130469776c1518355122d8d61577d4fefd6c8b))
* **event-aggregator:** removing another callback when called twice ([50e01cd7](http://github.com/aurelia/event-aggregator/commit/50e01cd7d882aaef49b7ff892b324ed4088e0b15))
* **index:**
  * update to latest framework configuration api ([87083a44](http://github.com/aurelia/event-aggregator/commit/87083a44abaf411f694460ff7308c17c43b682b3))
  * remove invalid character ([96642658](http://github.com/aurelia/event-aggregator/commit/96642658caf7d90733a0e103980841ccbbc43eb8))
* **package:** change jspm directories ([84c67fbc](http://github.com/aurelia/event-aggregator/commit/84c67fbc43971830255d20a08d726de9565b6dbd))
* **publish:** catch errors thrown in subscribers ([9dafe002](http://github.com/aurelia/event-aggregator/commit/9dafe002f1b2d8fbda5211495bb1de06c0b58c90))


#### Features

* **all:**
  * subscribe methods now return subscription object with dispose method ([862cb83f](http://github.com/aurelia/event-aggregator/commit/862cb83f2f1a3979fc39a90dbdbff4b1369dd252), closes [#10](http://github.com/aurelia/event-aggregator/issues/10))
  * add more type info ([7b08cae1](http://github.com/aurelia/event-aggregator/commit/7b08cae11ecc11c6e6130cfdfa9ded657c565ff4))
* **build:** update compiler and switch to register module format ([302c2870](http://github.com/aurelia/event-aggregator/commit/302c287092d9812a3d2c7e5fca2c1bb20618fffc))
* **docs:** generate api.json from .d.ts file ([36ac335d](http://github.com/aurelia/event-aggregator/commit/36ac335d10d43b02fb0c4402dcfceb8946a89222))
* **event-aggregator:**
  * subscribeOnce function ([076264e7](http://github.com/aurelia/event-aggregator/commit/076264e7ffc2bcec090b4c4339671262ffb8238d))
  * enable plugin usage ([2c597329](http://github.com/aurelia/event-aggregator/commit/2c5973292dbd5087325226298ae966b1dcb7ac17))


## 0.8.0 (2015-09-04)


#### Bug Fixes

* **build:** update linting, testing and tools ([172366a4](http://github.com/aurelia/event-aggregator/commit/172366a4e3ce6b97c16fc0f6af96e2e32114ba2e))


#### Features

* **docs:** generate api.json from .d.ts file ([36ac335d](http://github.com/aurelia/event-aggregator/commit/36ac335d10d43b02fb0c4402dcfceb8946a89222))


## 0.7.0 (2015-08-14)


#### Bug Fixes

* **index:** update to latest framework configuration api ([87083a44](http://github.com/aurelia/event-aggregator/commit/87083a44abaf411f694460ff7308c17c43b682b3))


#### Features

* **all:** add more type info ([7b08cae1](http://github.com/aurelia/event-aggregator/commit/7b08cae11ecc11c6e6130cfdfa9ded657c565ff4))


### 0.6.2 (2015-07-29)

* improve output file name

### 0.6.1 (2015-07-07)


## 0.6.0 (2015-07-02)


## 0.5.0 (2015-06-08)


#### Bug Fixes

* **publish:** catch errors thrown in subscribers ([9dafe002](http://github.com/aurelia/event-aggregator/commit/9dafe002f1b2d8fbda5211495bb1de06c0b58c90))


## 0.4.0 (2015-04-30)


#### Bug Fixes

* **all:** update to latest plugin api ([2a17472d](http://github.com/aurelia/event-aggregator/commit/2a17472dba61e6d6f364207d2ba438e65a97ed87))
* **event-aggregator:** removing another callback when called twice ([50e01cd7](http://github.com/aurelia/event-aggregator/commit/50e01cd7d882aaef49b7ff892b324ed4088e0b15))


## 0.3.0 (2015-04-09)


#### Bug Fixes

* **all:** upgrade to latest compiler ([ae84b313](http://github.com/aurelia/event-aggregator/commit/ae84b3139c1cc906614a904d78935e06a7959fde))


#### Features

* **event-aggregator:** subscribeOnce function ([076264e7](http://github.com/aurelia/event-aggregator/commit/076264e7ffc2bcec090b4c4339671262ffb8238d))


### 0.2.4 (2015-02-28)


#### Bug Fixes


test/setup.ts
let allTestFiles: string[] = [];
let allTemplatePaths: {[requirePath: string]: string} = {};
let TEST_REGEXP = /^\/base\/dist\/test\/test\/[^\/]+\.js$/i;
let TEMPLATE_REGEXP = /^\/base\/(src|test)\/.+\.html$/i;

interface Window {
  __karma__: any;
  require: any;
}

// Get a list of all the test files to include
Object.keys(window.__karma__.files).forEach(file => {
  if (TEST_REGEXP.test(file) && file !== '/base/dist/test/test/setup.js') {
    // Normalize paths to RequireJS module names.
    // If you require sub-dependencies of test files to be loaded as-is (requiring file extension)
    // then do not normalize the paths
    const normalizedTestModule = file.replace(/^\/base\/|\.js$/g, '');
    allTestFiles.push(normalizedTestModule);
  } else if (TEMPLATE_REGEXP.test(file)) {
    const normalizedTemplate = file.replace(/^\/base\//, '/base/dist/');
    allTemplatePaths[normalizedTemplate] = file;
  }
});

let started = false;

window.require.config({
  // Karma serves files under /base, which is the basePath from your config file
  baseUrl: '/base',

  deps: [],

  // we have to kickoff jasmine, as it is asynchronous
  callback: () => {
    if (started) {
      return;
    }
    started = true;
    window.require(allTestFiles, () => window.__karma__.start());
  },
  paths: {
    /* tslint:disable:max-line-length */
    'aurelia-binding': '/base/node_modules/aurelia-binding/dist/amd/aurelia-binding',
    'aurelia-bootstrapper': '/base/node_modules/aurelia-bootstrapper/dist/amd/aurelia-bootstrapper',
    'aurelia-dependency-injection': '/base/node_modules/aurelia-dependency-injection/dist/amd/aurelia-dependency-injection',
    'aurelia-event-aggregator': '/base/node_modules/aurelia-event-aggregator/dist/amd/aurelia-event-aggregator',
    'aurelia-framework': '/base/node_modules/aurelia-framework/dist/amd/aurelia-framework',
    'aurelia-history': '/base/node_modules/aurelia-history/dist/amd/aurelia-history',
    'aurelia-history-browser': '/base/node_modules/aurelia-history-browser/dist/amd/aurelia-history-browser',
    'aurelia-loader': '/base/node_modules/aurelia-loader/dist/amd/aurelia-loader',
    'aurelia-loader-default': '/base/node_modules/aurelia-loader-default/dist/amd/aurelia-loader-default',
    'aurelia-logging': '/base/node_modules/aurelia-logging/dist/amd/aurelia-logging',
    'aurelia-logging-console': '/base/node_modules/aurelia-logging-console/dist/amd/aurelia-logging-console',
    'aurelia-metadata': '/base/node_modules/aurelia-metadata/dist/amd/aurelia-metadata',
    'aurelia-pal': '/base/node_modules/aurelia-pal/dist/amd/aurelia-pal',
    'aurelia-pal-browser': '/base/node_modules/aurelia-pal-browser/dist/amd/aurelia-pal-browser',
    'aurelia-path': '/base/node_modules/aurelia-path/dist/amd/aurelia-path',
    'aurelia-polyfills': '/base/node_modules/aurelia-polyfills/dist/amd/aurelia-polyfills',
    'aurelia-router': '/base/node_modules/aurelia-router/dist/amd/aurelia-router',
    'aurelia-route-recognizer': '/base/node_modules/aurelia-route-recognizer/dist/amd/aurelia-route-recognizer',
    'aurelia-task-queue': '/base/node_modules/aurelia-task-queue/dist/amd/aurelia-task-queue',
    'aurelia-templating': '/base/node_modules/aurelia-templating/dist/amd/aurelia-templating',
    'aurelia-templating-binding': '/base/node_modules/aurelia-templating-binding/dist/amd/aurelia-templating-binding',
    /* tslint:enable:max-line-length */
    /* tslint:disable */
    'text': '/base/node_modules/requirejs-text/text',
    'template-registry-entry!dist/test/test/resources/my-component.html_unnormalized2': '/base/test/resources/my-component.html',
    /* tslint:enable */
    ...allTemplatePaths
  },
  packages: [
    {
      name: 'aurelia-event-aggregator',
      location: '/base/node_modules/aurelia-event-aggregator/dist/amd',
      main: 'aurelia-event-aggregator'
    }
  ]
});


test/jasmine.d.ts
declare var beforeAll: { (fn: () => void): void };
declare var beforeEach: { (fn: () => void): void };
declare var afterEach: { (fn: () => void): void };
declare var describe: { (name: string, fn: () => void): void };
declare var fail: { (reason: string): void };
declare var it: { (name: string, fn: (done: (failure?: any) => void) => void): void };
declare var expect: (x: any) => any;
declare var jasmine: { createSpy: () => any; };


test/event-aggregator.spec.js
import {EventAggregator} from '../src/index';

describe('event aggregator', () => {

  describe('subscribe', () => {

    describe('string events', () => {

      it('should not remove another callback when execute called twice', ()=>{
        let ea = new EventAggregator();
        let data = 0;

        let subscription = ea.subscribe('dinner', function() {});
        ea.subscribe('dinner', function() { data = 1; });

        subscription.dispose();
        subscription.dispose();

        ea.publish('dinner');

        expect(data).toBe(1);
      });

      it('adds event with callback to the eventLookup object', () =>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribe('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0]).toBe(callback);
      });

      it('adds multiple callbacks the same event', () =>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribe('dinner', callback);

        let callback2 = function(){};
        ea.subscribe('dinner', callback2);

        expect(ea.eventLookup.dinner.length).toBe(2);
        expect(ea.eventLookup.dinner[0]).toBe(callback);
        expect(ea.eventLookup.dinner[1]).toBe(callback2);
      });

      it('removes the callback after execution', ()=>{
        let ea = new EventAggregator();

        let callback = function(){};
        let subscription = ea.subscribe('dinner', callback);

        let callback2 = function(){};
        let subscription2 = ea.subscribe('dinner', callback2);

        expect(ea.eventLookup.dinner.length).toBe(2);
        expect(ea.eventLookup.dinner[0]).toBe(callback);
        expect(ea.eventLookup.dinner[1]).toBe(callback2);

        subscription.dispose();

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0]).toBe(callback2);

        subscription2.dispose();
        expect(ea.eventLookup.dinner.length).toBe(0);
      });

      it('will respond to an event any time it is published', ()=>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribe('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0]).toBe(callback);

        ea.publish('dinner');
        ea.publish('dinner');
        ea.publish('dinner');

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0]).toBe(callback);
      });

      it('will pass published data to the callback function', ()=>{
        let ea = new EventAggregator();
        let data = null;
        let callback = function(d){ data = d;};
        ea.subscribe('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0]).toBe(callback);

        ea.publish('dinner',{foo:"bar"});
        expect(data.foo).toBe("bar");
      });

    });

    describe('handler events', () =>{

      it('should not remove another handler when execute called twice', ()=>{
        let ea = new EventAggregator();
        let data = 0;

        let subscription = ea.subscribe(DinnerEvent, function() {});
        ea.subscribe(AnotherDinnerEvent, function() { data = 1; });

        subscription.dispose();
        subscription.dispose();

        ea.publish(new AnotherDinnerEvent(""));

        expect(data).toBe(1);
      });

      it('adds handler with messageType and callback to the messageHandlers array', ()=>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribe(DinnerEvent, callback);

        expect(ea.messageHandlers.length).toBe(1);
        expect(ea.messageHandlers[0].messageType).toBe(DinnerEvent);
        expect(ea.messageHandlers[0].callback).toBe(callback);
      });

      it('removes the handler after execution', () =>{
        let ea = new EventAggregator();
        let callback = function(){};
        let subscription = ea.subscribe(DinnerEvent, callback);

        expect(ea.messageHandlers.length).toBe(1);
        subscription.dispose();
        expect(ea.messageHandlers.length).toBe(0);
      });

    });

  });

  describe('subscribeOnce', () =>{

    describe('string events', () =>{

      it('adds event with an anynomous function that will execute the callback to the eventLookup object', ()=>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribeOnce('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);
      });

      it('adds multiple callbacks the same event', () =>{
        let ea = new EventAggregator();
        let callback = function(){};
        ea.subscribeOnce('dinner', callback);

        let callback2 = function(){};
        ea.subscribeOnce('dinner', callback2);

        expect(ea.eventLookup.dinner.length).toBe(2);
        expect(ea.eventLookup.dinner[0] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);
        expect(ea.eventLookup.dinner[1] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[1] === "function").toBe(true);
      });

      it('removes the callback after execution', ()=>{
        let ea = new EventAggregator();
        let callback = function(){};
        let subscription = ea.subscribeOnce('dinner', callback);

        let callback2 = function(){};
        let subscription2 = ea.subscribeOnce('dinner', callback2);

        expect(ea.eventLookup.dinner.length).toBe(2);
        expect(ea.eventLookup.dinner[0] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);
        expect(ea.eventLookup.dinner[1] === callback2).toBe(false);
        expect(typeof ea.eventLookup.dinner[1] === "function").toBe(true);

        subscription.dispose();

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);

        subscription2.dispose();
        expect(ea.eventLookup.dinner.length).toBe(0);
      });

      it('will respond to an event only once', ()=>{
        let ea = new EventAggregator();
        let data = null;

        let callback = function(){data = "something";};
        ea.subscribeOnce('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);

        ea.publish('dinner');
        expect(data).toBe("something");

        expect(ea.eventLookup.dinner.length).toBe(0);

        data = null;
        ea.publish('dinner');
        expect(data).toBe(null);
      });

      it('will pass published data to the callback function', ()=>{
        let ea = new EventAggregator();

        let data = null;
        let callback = function(d){data = d;};
        ea.subscribeOnce('dinner', callback);

        expect(ea.eventLookup.dinner.length).toBe(1);
        expect(ea.eventLookup.dinner[0] === callback).toBe(false);
        expect(typeof ea.eventLookup.dinner[0] === "function").toBe(true);

        ea.publish('dinner',{foo:"bar"});
        expect(data.foo).toBe("bar");

        data = null;
        ea.publish('dinner');
        expect(data).toBe(null);
      });
    });

    describe('handler events', () =>{

      it('adds handler with messageType and callback to the messageHandlers array', ()=>{
        let ea = new EventAggregator();

        let callback = function(){};
        ea.subscribeOnce(DinnerEvent, callback);

        expect(ea.messageHandlers.length).toBe(1);
        expect(ea.messageHandlers[0].messageType).toBe(DinnerEvent);
        expect(ea.messageHandlers[0].callback === callback).toBe(false);
        expect(typeof ea.messageHandlers[0].callback === "function").toBe(true);

      });

      it('removes the handler after execution', () =>{
        let ea = new EventAggregator();
        let callback = function(){};
        let subscription = ea.subscribeOnce(DinnerEvent, callback);

        expect(ea.messageHandlers.length).toBe(1);
        subscription.dispose();
        expect(ea.messageHandlers.length).toBe(0);
      });

    });

  });

  describe('publish', () =>{

    describe('string events', () =>{

      it('calls the callback functions for the event', () =>{
        let ea = new EventAggregator();

        let someData, someData2;

        let callback = function(data){
          someData = data;
        };
        ea.subscribe('dinner', callback);

        let callback2 = function(data){
          someData2 = data;
        };
        ea.subscribe('dinner', callback2);

        let data = {foo: 'bar'};
        ea.publish('dinner', data);

        expect(someData).toBe(data);
        expect(someData2).toBe(data);
      });

      it('does not call the callback functions if subscriber does not exist', () =>{
        let ea = new EventAggregator();

        let someData;

        let callback = function(data){
          someData = data;
        };
        ea.subscribe('dinner', callback);

        ea.publish('garbage', {});

        expect(someData).toBeUndefined();
      });

      it('handles errors in subscriber callbacks', () => {
        let ea = new EventAggregator();

        let someMessage;

        let crash = function() {
          throw new Error('oops');
        }

        let callback = function(message){
          someMessage = message;
        };

        let data = {foo: 'bar'};

        ea.subscribe('dinner', crash);
        ea.subscribe('dinner', callback);
        ea.subscribe('dinner', crash);

        ea.publish('dinner', data);

        expect(someMessage).toBe(data);
      });

    });

    describe('handler events', () =>{

      it('calls the callback functions for the event', () =>{
        let ea = new EventAggregator();

        let someMessage;

        let callback = function(message){
          someMessage = message;
        };
        ea.subscribe(DinnerEvent, callback);

        let americanDinner = new DinnerEvent('Cajun chicken');
        ea.publish(americanDinner);

        expect(someMessage.message).toBe('Cajun chicken');

        let swedishDinner = new DinnerEvent('Meatballs');
        ea.publish(swedishDinner);

        expect(someMessage.message).toBe('Meatballs');
      });

      it('does not call the callback funtions if message is not an instance of the messageType', ()=>{
        let ea = new EventAggregator();

        let someMessage;

        let callback = function(message){
          someMessage = message;
        };
        ea.subscribe(DinnerEvent, callback);

        ea.publish(new DrinkingEvent());

        expect(someMessage).toBeUndefined();
      });


      it('handles errors in subscriber callbacks', () => {
        let ea = new EventAggregator();

        let someMessage;

        let crash = function() {
          throw new Error('oops');
        }

        let callback = function(message){
          someMessage = message;
        };

        var data = {foo: 'bar'};

        ea.subscribe(DinnerEvent, crash);
        ea.subscribe(DinnerEvent, callback);

        ea.publish(new DinnerEvent(data));

        expect(someMessage.message).toBe(data);
      });

    });

  });

});

class DinnerEvent {
  constructor(message){
    this._message = message;
  }

  get message(){
    return this._message;
  }
}

class AnotherDinnerEvent {
  constructor(message){
    this._message = message;
  }

  get message(){
    return this._message;
  }
}

class DrinkingEvent {}


config.js
System.config({
  defaultJSExtensions: true,
  transpiler: "babel",
  babelOptions: {
    "optional": [
      "runtime",
      "optimisation.modules.system"
    ]
  },
  paths: {
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "aurelia-logging": "npm:aurelia-logging@1.0.0",
    "babel": "npm:babel-core@5.8.38",
    "babel-runtime": "npm:babel-runtime@5.8.38",
    "core-js": "npm:core-js@2.4.1",
    "github:jspm/nodelibs-assert@0.1.0": {
      "assert": "npm:assert@1.4.1"
    },
    "github:jspm/nodelibs-buffer@0.1.0": {
      "buffer": "npm:buffer@3.6.0"
    },
    "github:jspm/nodelibs-path@0.1.0": {
      "path-browserify": "npm:path-browserify@0.0.0"
    },
    "github:jspm/nodelibs-process@0.1.2": {
      "process": "npm:process@0.11.6"
    },
    "github:jspm/nodelibs-util@0.1.0": {
      "util": "npm:util@0.10.3"
    },
    "github:jspm/nodelibs-vm@0.1.0": {
      "vm-browserify": "npm:vm-browserify@0.0.4"
    },
    "npm:assert@1.4.1": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "buffer": "github:jspm/nodelibs-buffer@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "util": "npm:util@0.10.3"
    },
    "npm:babel-runtime@5.8.38": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:buffer@3.6.0": {
      "base64-js": "npm:base64-js@0.0.8",
      "child_process": "github:jspm/nodelibs-child_process@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "ieee754": "npm:ieee754@1.1.6",
      "isarray": "npm:isarray@1.0.0",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:core-js@2.4.1": {
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "path": "github:jspm/nodelibs-path@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "systemjs-json": "github:systemjs/plugin-json@0.1.2"
    },
    "npm:inherits@2.0.1": {
      "util": "github:jspm/nodelibs-util@0.1.0"
    },
    "npm:path-browserify@0.0.0": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:process@0.11.6": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "vm": "github:jspm/nodelibs-vm@0.1.0"
    },
    "npm:util@0.10.3": {
      "inherits": "npm:inherits@2.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:vm-browserify@0.0.4": {
      "indexof": "npm:indexof@0.0.1"
    }
  }
});


src/index.js
import * as LogManager from 'aurelia-logging';

const logger = LogManager.getLogger('event-aggregator');

class Handler {
  constructor(messageType, callback) {
    this.messageType = messageType;
    this.callback = callback;
  }

  handle(message) {
    if (message instanceof this.messageType) {
      this.callback.call(null, message);
    }
  }
}

function invokeCallback(callback, data, event) {
  try {
    callback(data, event);
  } catch (e) {
    logger.error(e);
  }
}

function invokeHandler(handler, data) {
  try {
    handler.handle(data);
  } catch (e) {
    logger.error(e);
  }
}

/**
* Represents a disposable subsciption to an EventAggregator event.
*/
interface Subscription {
  /**
  * Disposes the subscription.
  */
  dispose(): void;
}

/**
* Enables loosely coupled publish/subscribe messaging.
*/
export class EventAggregator {
  /**
  * Creates an instance of the EventAggregator class.
  */
  constructor() {
    this.eventLookup = {};
    this.messageHandlers = [];
  }

  /**
  * Publishes a message.
  * @param event The event or channel to publish to.
  * @param data The data to publish on the channel.
  */
  publish(event: string | any, data?: any): void {
    let subscribers;
    let i;

    if (!event) {
      throw new Error('Event was invalid.');
    }

    if (typeof event === 'string') {
      subscribers = this.eventLookup[event];
      if (subscribers) {
        subscribers = subscribers.slice();
        i = subscribers.length;

        while (i--) {
          invokeCallback(subscribers[i], data, event);
        }
      }
    } else {
      subscribers = this.messageHandlers.slice();
      i = subscribers.length;

      while (i--) {
        invokeHandler(subscribers[i], event);
      }
    }
  }

  /**
  * Subscribes to a message channel or message type.
  * @param event The event channel or event data type.
  * @param callback The callback to be invoked when the specified message is published.
  */
  subscribe(event: string | Function, callback: Function): Subscription {
    let handler;
    let subscribers;

    if (!event) {
      throw new Error('Event channel/type was invalid.');
    }

    if (typeof event === 'string') {
      handler = callback;
      subscribers = this.eventLookup[event] || (this.eventLookup[event] = []);
    } else {
      handler = new Handler(event, callback);
      subscribers = this.messageHandlers;
    }

    let idx = subscribers.push(handler);

    return {
      dispose() {
        subscribers.splice(idx, 1);
      }
    };
  }

  /**
  * Subscribes to a message channel or message type, then disposes the subscription automatically after the first message is received.
  * @param event The event channel or event data type.
  * @param callback The callback to be invoked when the specified message is published.
  */
  subscribeOnce(event: string | Function, callback: Function): Subscription {
    let sub = this.subscribe(event, (a, b) => {
      sub.dispose();
      return callback(a, b);
    });

    return sub;
  }
}

/**
* Includes EA functionality into an object instance.
* @param obj The object to mix Event Aggregator functionality into.
*/
export function includeEventsIn(obj: Object): EventAggregator {
  let ea = new EventAggregator();

  obj.subscribeOnce = function(event, callback) {
    return ea.subscribeOnce(event, callback);
  };

  obj.subscribe = function(event, callback) {
    return ea.subscribe(event, callback);
  };

  obj.publish = function(event, data) {
    ea.publish(event, data);
  };

  return ea;
}

/**
* Configures a global EA by merging functionality into the Aurelia instance.
* @param config The Aurelia Framework configuration object used to configure the plugin.
*/
export function configure(config: Object): void {
  config.instance(EventAggregator, includeEventsIn(config.aurelia));
}


karma.conf.js
// Karma configuration
// Generated on Sun Aug 28 2016 19:03:27 GMT-0400 (Eastern Daylight Time)

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine', 'requirejs'],

    // list of files / patterns to load in the browser
    files: [
      'dist/test/test/setup.js',
      { pattern: 'dist/test/**/*.js', included: false, watched: true },
      { pattern: 'node_modules/**/*.js', included: false, watched: false },
    ],


    // list of files to exclude
    exclude: [],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}
